atomic_reactor/core.py

```diff
@@ -214,8 +215,56 @@ def build_image_privileged_container(self, build_image, json_args_path):
         return container_id
 
 
+def retry(function, *args, **kwargs):
+    retry_times = kwargs.pop('retry', None)
+    retry_delay = DOCKER_BACKOFF_FACTOR
+    retry_client_statuses = DOCKER_CLIENT_STATUS_RETRY
+
+    final_excep = None
+
+    for counter in range(int(retry_times)):
+        if counter > 0:
+            time.sleep(retry_delay * (2 ** counter))
+            final_excep = None
+        try:
+            return function(*args, **kwargs)
+        except APIError as e:
+            if e.response.status_code in retry_client_statuses:
+                logger.info("retrying %s on %s", function, e.response.status_code)
+                final_excep = e
+            else:
+                raise
+
+    if final_excep is not None:
+        raise final_excep
+
+
+class WrappedDocker(object):
+    def __init__(self, **kwargs):
+        self.retry_times = kwargs.pop('retry')
+        if not self.retry_times:
+            self.retry_times = 1
+        try:
+            # docker-py 2.x
+            self.wrapped = docker.APIClient(**kwargs)
+        except AttributeError:
+            # docker-py 1.x
+            self.wrapped = docker.Client(**kwargs)
+
+    def __getattr__(self, attr):
+        orig_attr = getattr(self.wrapped, attr)
+
+        if callable(orig_attr):
+            def hooked(*args, **kwargs):
+                return retry(orig_attr, *args, retry=self.retry_times, **kwargs)
```

Comment:
What if `retry` is given in the `kwargs` of the original function?
I believe python will complain about two values being given to the `retry` parameter.
Maybe this?
```python
kwargs.setdefault('retry', self.retry_times)
```
That way it gives the flexibility to overwrite the retry_times value for a specific call.

Reply:
no need anymore

Reply:
if any argument which isn't accepted with docker.APIClient / Client will be given in kwargs
init will complain:
TypeError: __init__() got an unexpected keyword argument 'retry'

Reply:
That's not what I mean. If `orig_attr` is, let's say, `build` function of docker-py's `Client` class and it takes a `retry` parameter, we won't be able to use it.

AFAIK, there are no methods there that take a `retry` parameter so this is probably fine. But it's the kind of thing that increases difficulty in debugging.

Reply:
yeah docker method can't have 'retry' parameter, as we have no way passing 2 same named parameters to that new class, unless I'd make extra method just to set retry